### PR TITLE
(PUP-7875) Use bolt to execute tasks in run_task

### PR DESCRIPTION
**Pull request by pcarlisle**
Run tasks using the bolt executor

### CODE REVIEW

File: lib/puppet/functions/run_task.rb [hunk 1]
```diff
@@ -21,23 +21,46 @@ def mocked_run_task(task, *hosts)
         {:operation => 'run_task'})
     end
 
+    begin
+      require 'bolt/executor'
+      require 'bolt/node'
+    rescue LoadError
+      raise Puppet::ParseErrorWithIssue.from_issue_and_stack(Puppet::Pops::Issues::TASK_MISSING_BOLT)
+    end
```
Comment (hlindberg) [inline lib/puppet/functions/run_task.rb]: The above is best done with a feature - see how this is done in "eyaml_lookup_key.rb" 
```
  unless Puppet.features.hiera_eyaml?
    raise Puppet::DataBinding::LookupError, 'Lookup using eyaml lookup_key function is only supported when the hiera_eyaml library is present'
  end
```

Comment (hlindberg) [reply lib/puppet/functions/run_task.rb]: and that snippet is placed before the call to dispatch

File: lib/puppet/functions/run_task.rb [hunk 2]
```diff
@@ -21,23 +21,46 @@ def mocked_run_task(task, *hosts)
...
+
+    results.map do |host, result|
+      output = StringIO.new()
+      result.print_to_stream(output)
+      call_function('notice', "Ran task #{task._pcore_type.name} on '#{host}' with result '#{output.string.chomp}'")
+      if result.success?
+        output.string
       else
-        exit_code
+        result.exit_code
```
Comment (hlindberg) [inline lib/puppet/functions/run_task.rb]: The produced resulting hash is fine for now, but something to talk more about.

File: lib/puppet/functions/run_task.rb [hunk 3]
```diff
@@ -21,23 +21,46 @@ def mocked_run_task(task, *hosts)
...
-      exit_code = task.respond_to?(:simulated_exit_code) ? task.exit_code : 0
-      result =
-      if exit_code == 0
-        result = task.respond_to?(:simluated_result) ? task.simulated_result : '<simulated result>'
+    executor = Bolt::Executor.new(hosts.map { |host| Bolt::Node.from_uri(host, nil, nil, nil) })
+
+    # TODO: separate handling of default since it's platform specific
+    input_method = task._pcore_type['input_method'].value
+
+    # Should have a uniform way to retrieve arguments
```
Comment (hlindberg) [inline lib/puppet/functions/run_task.rb]: Yeah, don't know about this - ping @thallgren

**Comment by Iristyle** [comment]
For the sake of documentation, https://github.com/pcarlisle/puppet/blob/run-task-with-bolt/lib/puppet/pops/types/task.rb holds the definition of the task object, which is the `Object` parameter supplied in the `dispatcher` block.

File: lib/puppet/functions/run_task.rb [hunk 4]
```diff
@@ -9,35 +9,53 @@
...
+                  task._pcore_init_hash
+                end
+    raw_results = executor.run_task(task.executable_path, input_method, arguments)
+
+    results = {}
+    raw_results.each do |node, result|
+      results[node.uri] = result
+    end
+
+    results.map do |host, result|
```
Comment (lutter) [inline lib/puppet/functions/run_task.rb]: I think we need to produce a more finegrained result: for each node, we want to preserve `stdout`, `stderr`, and `exit_code` for inspection in the Puppet code.

We'll also want to parse JSON output if the task returns that, though that can happen in another PR.

Comment (hlindberg) [reply lib/puppet/functions/run_task.rb]: Lets add specific tickets for that. We also have one for result aggregation when running a plan.

File: lib/puppet/functions/run_task.rb [hunk 5]
```diff
@@ -9,35 +9,53 @@
...
+      output = result.output_string
+      call_function('notice', "Ran task #{task._pcore_type.name} on '#{host}' with result '#{output.chomp}'")
```
Comment (hlindberg) [inline lib/puppet/functions/run_task.rb]: Don't think this output is wanted unless --debug is on

File: spec/unit/functions/run_task_spec.rb [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+    it 'when running a task with metadata' do
+      executor = mock('executor')
+      executable = File.join(env_dir, 'modules/test/tasks/meta.sh')
+
+      Bolt::Executor.expects(:from_uris).with(hosts).returns(executor)
+      executor.expects(:run_task).with(executable, 'environment', {'message' => 'the message'}).returns({ host => result })
+
+      expect(eval_and_collect_notices(<<-CODE, node)).to eql(["Ran task Test::Meta on '#{hostname}' with result '#{message}'"])
+        run_task(Test::Meta({message => "#{message}"}), "#{hostname}")
```
Comment (hlindberg) [inline spec/unit/functions/run_task_spec.rb]: Instead of checking that it logged the result, it should check that it returns the result.

**Comment by hlindberg** [comment]
There are more nips and tucks needed, but this is fine as a starting point. Merging.

**Pull request merged**